modules/auxiliary/gather/ldap_esc_vulnerable_cert_finder.rb

```diff
@@ -0,0 +1,544 @@
+    fail_with(Failure::NotVulnerable, 'Could not find any ESC3 template 2 vulnerable certificate templates!') if esc3_template_2_entries.blank?
+
+    # Grab a list of certificates that contain vulnerable settings.
+    # Also print out the list of SIDs that can enroll in that server.
+
+    esc3_template_2_entries.each do |entry|
+      @vuln_certificate_template_list << entry[:cn][0]
+
+      begin
+        security_descriptor = Rex::Proto::MsDtyp::MsDtypSecurityDescriptor.read(entry[:ntsecuritydescriptor][0])
+      rescue IOError => e
+        fail_with(Failure::UnexpectedReply, "Unable to read security descriptor! Error was: #{e.message}")
+      end
+
+      flag_allowed_to_enroll, allowed_sids = parse_dacl_or_sacl(security_descriptor.dacl) if security_descriptor.dacl
+      if flag_allowed_to_enroll == true
+        @vuln_cert_template_table << ['ESC3_TEMPLATE_2', entry[:cn][0], entry[:dn][0], convert_sids_to_human_readable_name(allowed_sids)]
+      end
+      parse_dacl_or_sacl(security_descriptor.sacl) if security_descriptor.sacl
+    end
+  end
+
+  def find_enrollable_vuln_certificate_templates
+    # For each of the vulnerable certificate templates, determine which servers
+    # allows users to enroll in that certificate template and which users/groups
+    # have permissions to enroll in certificates on each server.
+
+    # First remove duplicate certificate templates. Aka certificate templates that may
+    # be vulnerable to multiple attack vectors. This will ensure we only have one copy
+    # of each vulnerable certificate template instead of multiple copies.
+    @vuln_certificate_template_list.uniq!
+
+    @vuln_certificate_template_list.each do |certificate_template|
+      certificate_enrollment_raw_filter = "(&(objectClass=pKIEnrollmentService)(certificateTemplates=#{certificate_template}))"
+      attributes = ['dnsHostname', 'ntsecuritydescriptor']
+      base_prefix = 'CN=Enrollment Services,CN=Public Key Services,CN=Services,CN=Configuration'
+      enrollment_ca_data = query_ldap_server(certificate_enrollment_raw_filter, attributes, base_prefix: base_prefix)
+      next if enrollment_ca_data.blank?
+
+      enrollment_ca_data.each do |ca_server|
+        begin
+          security_descriptor = Rex::Proto::MsDtyp::MsDtypSecurityDescriptor.read(ca_server[:ntsecuritydescriptor][0])
+        rescue IOError => e
+          fail_with(Failure::UnexpectedReply, "Unable to read security descriptor! Error was: #{e.message}")
+        end
+
+        flag_allowed_to_enroll, allowed_sids = parse_dacl_or_sacl(security_descriptor.dacl) if security_descriptor.dacl
+        if flag_allowed_to_enroll == true
+          @enrollment_allowed_table << [ca_server[:dnshostname][0], certificate_template, convert_sids_to_human_readable_name(allowed_sids)]
+        end
+        parse_dacl_or_sacl(security_descriptor.sacl) if security_descriptor.sacl
```

Comment:
Same comment again about this block. flag_allowed_to_enroll is buggy, it should just be if flag_allowed_to_enroll and then parse_dacl_or_sacl doesn't appear to do anything.

Reply:
Fixed this up with https://github.com/rapid7/metasploit-framework/commit/fd33cdc58e47576823a874c5f17d81b3e5660f9b which uses the new function you suggested and also fixes the aforementioned issues.